Client/core/Graphics/CMaterialPrimitive3DBatcher.cpp

```diff
@@ -0,0 +1,243 @@
+        const void* pVertexStreamZeroData = &primitive.pVecVertices->at(0);
+        size_t iCollectionSize = primitive.pVecVertices->size();
+
+        CMaterialItem* pMaterial = primitive.pMaterial;
+        if (pMaterial != pLastMaterial)
+        {
+            // Set texture addressing mode
+            m_pDevice->SetSamplerState(0, D3DSAMP_ADDRESSU, pMaterial->m_TextureAddress);
+            m_pDevice->SetSamplerState(0, D3DSAMP_ADDRESSV, pMaterial->m_TextureAddress);
+
+            if (pMaterial->m_TextureAddress == TADDRESS_BORDER)
+                m_pDevice->SetSamplerState(0, D3DSAMP_BORDERCOLOR, pMaterial->m_uiBorderColor);
+        }
+
+        if (CTextureItem* pTextureItem = DynamicCast<CTextureItem>(pMaterial))
+        {
+            m_pDevice->SetTexture(0, pTextureItem->m_pD3DTexture);
+            DrawPrimitive(primitive.eType, primitive.pVecVertices->size(), pVertexStreamZeroData, uiVertexStreamZeroStride);
+        }
+        else if (CShaderInstance* pShaderInstance = DynamicCast<CShaderInstance>(pMaterial))
+        {
+            // Draw using shader
+            ID3DXEffect* pD3DEffect = pShaderInstance->m_pEffectWrap->m_pD3DEffect;
+
+            if (pMaterial != pLastMaterial)
+            {
+                // Apply custom parameters
+                pShaderInstance->ApplyShaderParameters();
+                // Apply common parameters
+                pShaderInstance->m_pEffectWrap->ApplyCommonHandles();
+                // Apply mapped parameters
+                pShaderInstance->m_pEffectWrap->ApplyMappedHandles();
+            }
+
+            // Do shader passes
+            DWORD dwFlags = D3DXFX_DONOTSAVESHADERSTATE;
+            uint  uiNumPasses = 0;
+            pShaderInstance->m_pEffectWrap->Begin(&uiNumPasses, dwFlags, false);
+
+            for (uint uiPass = 0; uiPass < uiNumPasses; uiPass++)
+            {
+                pD3DEffect->BeginPass(uiPass);
+                DrawPrimitive(primitive.eType, primitive.pVecVertices->size(), pVertexStreamZeroData, uiVertexStreamZeroStride);
+                pD3DEffect->EndPass();
+            }
+            pShaderInstance->m_pEffectWrap->End();
+
+            // If we didn't get the effect to save the shader state, clear some things here
+            if (dwFlags & D3DXFX_DONOTSAVESHADERSTATE)
+            {
+                m_pDevice->SetVertexShader(NULL);
+                m_pDevice->SetPixelShader(NULL);
+            }
+        }
+        pLastMaterial = pMaterial;
+        pMaterial->Release();
+    }
+
+    // Clean up
+    ClearQueue();
+
+    if (pSavedStateBlock)
+    {
+        pSavedStateBlock->Apply();
+        SAFE_RELEASE(pSavedStateBlock);
+    }
+}
+
+void CMaterialPrimitive3DBatcher::ClearQueue()
+{
+    // Clean up
+    for (auto& primitive : m_primitiveList)
+    {
+        delete primitive.pVecVertices;
+    }
+
+    size_t prevSize = m_primitiveList.size();
+    m_primitiveList.clear();
+    m_primitiveList.reserve(prevSize);
```

Comment:
Remove this and the `size_t prevSize = m_primitiveList.size();` line.
`std::vector<T>::clear()` doesn't release the allocated memory, you don't have to call `std::vector<T>::reserve(size_t)`.

Reply:
@CrosRoad95 Exists still in latest version?